admin/code/AddToCampaignHandler.php

```diff
@@ -0,0 +1,232 @@
+<?php
+
+
+/**
+ * Class AddToCampaignHandler - handle the AddToCampaign action.
+ *
+ * This is a class designed to be delegated to by a Form action handler method in the EditForm of a LeftAndMain
+ * child class.
+ *
+ * Add To Campaign can be seen as an item action like "publish" or "rollback", but unlike those actions
+ * it needs one additional piece of information to execute, the ChangeSet ID.
+ *
+ * So this handler does one of two things to respond to the action request, depending on whether the ChangeSet ID
+ * was included in the submitted data
+ * - If it was, perform the Add To Campaign action (as per any other action)
+ * - If it wasn't, return a form to get the ChangeSet ID and then repeat this action submission
+ *
+ * To use, you'd add an action to your LeftAndMain subclass, like this:
+ *
+ *     function addtocampaign($data, $form) {
+ *         $handler = AddToCampaignHandler::create($form, $data);
+ *         return $handler->handle();
+ *     }
+ *
+ *  and add an AddToCampaignHandler_FormAction to the EditForm, possibly through getCMSActions
+ */
+class AddToCampaignHandler {
+
+	/**
+	 * @var Form - the EditForm that contains the action we're being delegated to from
+	 */
+	protected $editForm;
+	/**
+	 * @var [string] - the submitted form data
```

Comment:
Sorry, array of strings is `@var string[]` :P

Its also not an array of explicitly strings; It can be nested arrays. You probably just want to cast this as `@var array`